### PR TITLE
Improve docker setup & config

### PR DESCRIPTION
**Pull request by Pwuts**
### Background
The current Dockerfile and docker-compose file seem to have unnecessary convolution of configuration and don't contain an optimal configuration out of the box. Also, it is not consistent with the README. This PR attempts to fix that.

This PR either conflicts with or deprecates PRs #1830 and #1199

### Changes
* Merge `apt-get` actions into one layer
* Uncouple app user and python dependencies: install packages globally instead of in `/home/appuser`
* Create `appuser` with UID 1000 to match probable permissions on Linux host
* Change workdir to `/app` (as documented in the README) which is a widely used standard, making it easier to work with for new users
* Improve/fix `docker-compose` config
  * Use Redis container as memory backend by default
  * Remove unused (and after this PR, redundant) volume mount to `/app`
  * Mount `./logs/`, `./auto_gpt_workspace/` and `./ai_settings.yaml` in `/app`

### Documentation
New configuration is consistent with current configuration; no update needed.

### Test Plan
No Python code is changed, so testing succeeds if the container builds and runs properly after these changes, which it does.

### PR Quality Checklist
- [ ] My pull request is atomic and focuses on a single change. (as far as I understand this checkbox)
- [x] I have thoroughly tested my changes with multiple different prompts.
- [x] I have considered potential risks and mitigations for my changes.
- [x] I have documented my changes clearly and comprehensively.
- [x] I have not snuck in any "extra" small tweaks changes

### CODE REVIEW

File: Dockerfile [hunk 1]
```diff
@@ -1,26 +1,25 @@
...
-USER appuser
+# Install the required python packages globally
+ENV PATH="$PATH:/root/.local/bin"
+COPY requirements-docker.txt .
+RUN pip install --no-cache-dir -r requirements-docker.txt
 
-# Copy the requirements.txt file and install the requirements
-COPY --chown=appuser:appuser requirements-docker.txt .
-RUN pip install --no-cache-dir --user -r requirements-docker.txt
+# Create a non-root user
```
Comment (nponeccop) [inline Dockerfile]: Isn't appuser already in the base container?

**Review by nponeccop — APPROVED** [review]

File: Dockerfile (hunk 1 above)
Comment (Pwuts) [reply Dockerfile]: Not as far as I can find, no.

**Comment by ntindle** [comment]
May be worth adding 
Chrome for selenium in docker to the compose file

https://hub.docker.com/r/selenium/standalone-chrome/

See here for examples

https://github.com/SeleniumHQ/docker-selenium

**Comment by Pwuts** [comment]
@ntindle currently, Chrome and Firefox along with some other tools are installed in the docker container, see Dockerfile. I'm not convinced this is *the way*, but also don't have time to dive into it further atm.

Do you think it is worth it to use standalone selenium containers instead of running headless browsers in the `auto-gpt` container?

**Comment by ntindle** [comment]
IMO using the selenium headless versions with their drivers makes a lot of sense. Otherwise we’d basically need to duplicate the work they’ve done and will continue to do when things change

**Comment by Pwuts** [comment]
Could you make a PR to implement that as soon as this one merges? Would be nice :)

**Comment by ntindle** [comment]
Can do

**Comment by Pwuts** [comment]
@mikekelly care to review?

File: Dockerfile [hunk 2]
```diff
@@ -20,18 +20,18 @@ ENV PIP_NO_CACHE_DIR=yes \
...
-COPY --chown=appuser:appuser requirements.txt .
+# Install the required python packages globally
+ENV PATH="$PATH:/root/.local/bin"
+COPY requirements.txt .
 RUN sed -i '/Items below this point will not be included in the Docker Image/,$d' requirements.txt && \
-	pip install --no-cache-dir --user -r requirements.txt
+	pip install --no-cache-dir -r requirements.txt
+
+# Create a non-root user
+RUN useradd -u 1000 --create-home appuser
```
Comment (mikekelly) [inline Dockerfile]: there was a user who this was causing folder permissions issues for, so we reverted it back to simply using root.

File: docker-compose.yml [hunk 3]
```diff
@@ -9,9 +9,13 @@ services:
     build: ./
     env_file:
       - .env
+    environment:
```
Comment (mikekelly) [inline docker-compose.yml:12]: I think it's better if we just provide one location for setting the env variables. This is going to catch someone out

File: docker-compose.yml [hunk 4]
```diff
@@ -9,9 +9,13 @@ services:
...
+      MEMORY_BACKEND: ${MEMORY_BACKEND:-redis}
+      REDIS_HOST: ${REDIS_HOST:-redis}
     volumes:
-      - "./autogpt:/app"
-      - ".env:/app/.env"
+      - ./logs:/app/logs
```
Comment (mikekelly) [inline docker-compose.yml]: This is resolved in the orther PR I mentioned. We just copy and mount the entire project so the dir structure is completely consistent with running it outside docker

File: Dockerfile [hunk 5]
```diff
@@ -20,18 +20,18 @@ ENV PIP_NO_CACHE_DIR=yes \
...
-# Create a non-root user and set permissions
-RUN useradd --create-home appuser
-WORKDIR /home/appuser
-RUN chown appuser:appuser /home/appuser
-USER appuser
-
-# Copy the requirements.txt file and install the requirements
-COPY --chown=appuser:appuser requirements.txt .
+# Install the required python packages globally
+ENV PATH="$PATH:/root/.local/bin"
```
Comment (mikekelly) [inline Dockerfile:21]: Why was this necessary?

File: Dockerfile [hunk 6]
```diff
@@ -1,13 +1,13 @@
...
 
-# Install git
-RUN apt-get -y update
-RUN apt-get -y install git chromium-driver
-
-# Install Xvfb and other dependencies for headless browser testing
-RUN apt-get update \
-    && apt-get install -y wget gnupg2 libgtk-3-0 libdbus-glib-1-2 dbus-x11 xvfb ca-certificates
+# Install packages
+RUN apt-get update && apt-get install -y \
```
Comment (mikekelly) [inline Dockerfile:8]: I agree this should be cleaned up but you could split the `apt-get update` from the `apt-get install` in order to make optimal use of the docker cache when building

File: Dockerfile [hunk 7]
```diff
@@ -1,13 +1,13 @@
...
+    curl jq \
+    git \
+    chromium-driver \
+    # dependencies for headless browser testing
```
Comment (mikekelly) [inline Dockerfile]: consider just removing this comment about browser testing

File: Dockerfile (hunk 5 above)
Comment (mikekelly) [reply Dockerfile:21]: all of the changes in this file from this point down conflict with https://github.com/Significant-Gravitas/Auto-GPT/pull/1199

**Comment by mikekelly** [comment]
I don't see the purpose in creating a PR that "deprecates" a PR that precedes it, why not simply provide feedback on an existing PR that is waiting to be merged?

**Review by mikekelly — CHANGES_REQUESTED** [review]
This should be reduced in scope to propose introducing the environment configuration. Everything else is covered in #1199

File: Dockerfile (hunk 5 above)
Comment (Pwuts) [reply Dockerfile:21]: It was necessary to let `python` find globally installed modules regardless of workdir.

**Comment by Pwuts** [comment]
> I don't see the purpose in creating a PR that "deprecates" a PR that precedes it, why not simply provide feedback on an existing PR that is waiting to be merged?

Because I didn't know about your PR, and submitted it before becoming as actively involved as I am now. We're working on combing through *all* open PRs, but I hadn't gotten to this one yet.

**Comment by Pwuts** [comment]
Closing in favor of #1199

**Comment by Pwuts** [comment]
Reopening since #1199 isn't being updated

File: Dockerfile (hunk 6 above)
Comment (Pwuts) [reply Dockerfile:8]: The only thing that will trigger a rebuild of this layer is pulling a new version of `python:3.10-slim`. Considering that, it doesn't matter whether `update` and `install` are combined or executed in separate steps, right?

**Review by ntindle — APPROVED** [review]

**Comment by ntindle** [comment]
This combined with a smoke test of the docker container in CI would be top notch

**Comment by Pwuts** [comment]
@ntindle this PR fixes docker; a draft for a smoke test job is in #3059

**Review by ntindle — DISMISSED** [review]

**Comment by Pwuts** [comment]
I want to improve the instructions for reusing a container before merging

File: autogpt/commands/web_selenium.py [hunk 8]
```diff
@@ -93,8 +93,13 @@ def scrape_text_with_selenium(url: str) -> tuple[WebDriver, str]:
             options.add_argument("--headless")
             options.add_argument("--disable-gpu")
 
+        chromium_driver_path = Path("/usr/bin/chromedriver")
```
Comment (ntindle) [inline autogpt/commands/web_selenium.py:96]: Worried about cross platform of this line

Comment (Pwuts) [reply autogpt/commands/web_selenium.py:96]: oop

Comment (Pwuts) [reply autogpt/commands/web_selenium.py:96]: It doesn't break on Windows, just doesn't find the `chromedriver` and downloads it instead (`ChromeDriverManager().install()`)

File: docs/installation.md [hunk 9]
```diff
@@ -61,38 +52,54 @@ Note: If you don't have Git, you can just download the [latest stable release](h
...
+        ``` shell
+        # Replace string in angled brackets (<>) to your own ID
+        azure_model_map:
+            fast_llm_model_deployment_id: "<my-fast-llm-deployment-id>"
+            ...
+        ```
      - Details can be found here: https://pypi.org/project/openai/ in the `Microsoft Azure Endpoints` section and here: https://learn.microsoft.com/en-us/azure/cognitive-services/openai/tutorials/embeddings?tabs=command-line for the embedding model.
      - If you're on Windows you may need to install https://learn.microsoft.com/en-us/cpp/windows/latest-supported-vc-redist?view=msvc-170
 
-## Docker
```
Comment (bobinson) [inline docs/installation.md:75]: IMHO we can move docker related documentation to a separate file as in future it will see more inputs and we can provide a path for the same.

**Review by p-i- — APPROVED** [review]

File: docs/installation.md (hunk 9 above)
Comment (Pwuts) [reply docs/installation.md:75]: We'll be shifting stuff around in the docs anyways as we are looking to push docker as the preferred way of running Auto-GPT.
